Exposing variant classification methods in GenotypeConcordance.

nh13 opened the pull request:
`GenotypeConcordance` is a very useful tool to assess the sensitivity and specificity (and PPV) of variant calls.  Nonetheless, I would like to access some methods in `GenotypeConcordance` in tools that filter and inspect variant calls relative to truth.   As always, I am open to a change in naming as I am admittedly not good at that and would like constructive criticism to get better.

`GenotypeConcordance` methods that are exposed include `determineState` and `classifyVariants`.  I did not refactor `determineState` to use `Optional`s as it relies on some heavy "is something `null`" logic.  I also made the `PairedVariantSubContextIterator` public so that others can iterate through variant contexts similar to `GenotypeConcordance`.   I did not refactor the tests to test `classifyVariants` directly, as `classifyVariants` calls `determineState`, with the latter well tested, and `GenotypeConcordanceTest` does a good job of testing the variant counting which is what remains in `classifyVariants` after `determineState` is called.

Thanks for the review!

yfarjoun commented:
not a review yet, but the ipr slipped in.

On Tue, Apr 12, 2016 at 2:07 PM, Nils Homer <notifications@github.com>
wrote:

> Background
>
> GenotypeConcordance is a very useful tool to assess the sensitivity and
> specificity (and PPV) of variant calls. Nonetheless, I would like to access
> some methods in GenotypeConcordance in tools that filter and inspect
> variant calls relative to truth. As always, I am open to a change in naming
> as I am admittedly not good at that and would like constructive criticism
> to get better.
>
> GenotypeConcordance methods that are exposed include determineState and
> classifyVariants. I did not refactor determineState to use Optionals as
> it relies on some heavy "is something null" logic. I also made the
> PairedVariantSubContextIterator public so that others can iterate through
> variant contexts similar to GenotypeConcordance. I did not refactor the
> tests to test classifyVariants directly, as classifyVariants calls
> determineState, with the latter well tested, and GenotypeConcordanceTest
> does a good job of testing the variant counting which is what remains in
> classifyVariants after determineState is called.
>
> Thanks for the review!
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/broadinstitute/picard/pull/506
> Commit Summary
>
>    - Exposing variant classification methods in GenotypeConcordance.
>
> File Changes
>
>    - *M* Picard-public.ipr
>    <https://github.com/broadinstitute/picard/pull/506/files#diff-0> (67)
>    - *M* src/java/picard/vcf/GenotypeConcordance.java
>    <https://github.com/broadinstitute/picard/pull/506/files#diff-1> (212)
>    - *A* src/java/picard/vcf/PairedVariantSubContextIterator.java
>    <https://github.com/broadinstitute/picard/pull/506/files#diff-2> (119)
>    - *M* src/tests/java/picard/vcf/GenotypeConcordanceTest.java
>    <https://github.com/broadinstitute/picard/pull/506/files#diff-3> (6)
>
> Patch Links:
>
>    - https://github.com/broadinstitute/picard/pull/506.patch
>    - https://github.com/broadinstitute/picard/pull/506.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/broadinstitute/picard/pull/506>
>

eitanbanks commented on src/java/picard/vcf/GenotypeConcordance.java:
```diff
@@ -286,59 +286,22 @@ private boolean indexExists(final File vcf) {
+            final boolean stateClassified = classifyVariants(tuple.leftVariantContext, TRUTH_SAMPLE,
+                    tuple.rightVariantContext, CALL_SAMPLE,
+                    Optional.of(snpCounter), Optional.of(indelCounter),
+                    MIN_GQ, MIN_DP);
```
Much cleaner.  I like it.

eitanbanks commented on src/java/picard/vcf/GenotypeConcordance.java:
```diff
@@ -380,6 +343,82 @@ else if (callVariantContextType == INDEL) {
+
+    /**
+     * Attempts to determine the concordance state given the truth and all variant context.  Optionally increments the genotype concordance
+     * counts for SNPs and Indels respectively.  This will ignore cases where an indel was found the truth and SNP  was found  in the call,
```
Extra space: "SNP  was" (!!!)

eitanbanks replied:
Also, the grammer is a little messed up.  Please fix.
It will only update the count for a snp or an indel but not both, correct?

eitanbanks commented on src/java/picard/vcf/GenotypeConcordance.java:
```diff
@@ -380,6 +343,82 @@ else if (callVariantContextType == INDEL) {
+     * @param callContext A variant context representing the call
+     * @param callSample The name of the call sample
+     * @param snpCounter optionally a place to increment the counts for SNP truth/call states
+     * @param indelCounter optionally a place to increment the counts for SNP truth/call states
```
indel, not snp

eitanbanks commented on src/java/picard/vcf/GenotypeConcordance.java:
```diff
@@ -380,6 +343,82 @@ else if (callVariantContextType == INDEL) {
+        final TruthAndCallStates truthAndCallStates = determineState(truthContext.orElse(null), truthSample, callContext.orElse(null), callSample, minGq, minDp);
+        if (truthVariantContextType == SNP) {
+            if ((callVariantContextType == SNP) || (callVariantContextType == MIXED) || (callVariantContextType == NO_VARIATION)) {
+                // Note.  If truth is SNP and call is MIXED, the event will be logged in the indelCounter, with row = MIXED
```
Meaning that it will *also* be logged there?  Because it's being logged here in the snpCounter.

nh13 replied:
These were wrong before this PR, so I will fix them to what I think they *should* be.

eitanbanks commented on src/java/picard/vcf/PairedVariantSubContextIterator.java:
```diff
@@ -0,0 +1,119 @@
+
+        VcfTuple(final Optional<VariantContext> leftVariantContext, final VariantContext rightVariantContext) {
+            this(leftVariantContext, Optional.of(rightVariantContext));
+
```
extra space

eitanbanks commented on src/java/picard/vcf/PairedVariantSubContextIterator.java:
```diff
@@ -0,0 +1,119 @@
+        }
+    }
+
+    @Override public void remove() {
```
Should the override annotation be on a separate line (not sure of the convention in this codebase)?

nh13 replied:
Good point, I copied an pasted, will fix.

eitanbanks commented:
Comments made.  All fairly minor.  Looks good otherwise.

nh13 commented:
@eitanbanks responded to your comments.  Let me know if you would like me to rebase and squash, or if you want to use that fancy github merge button to do it for you.

eitanbanks commented:
What fancy github merge button?  Or am I just being lame and not getting it?

tfenne commented:
@nh13: The "fancy button" only shows up if you have write access in both the source and destination repos (so @eitanbanks won't see it on this PR).  @eitanbanks it allows you to squash during the merge on github if you have appropriate access.

eitanbanks commented:
Oh, that *is* fancy!  (And I'm also relieved that there wasn't some private joke I was missing.  :)  )

nh13 commented:
@eitanbanks TIL from @tfenne.  I have squashed and rebased onto master so you should be able to merge once the test pass.  Thanks for the fast review!
